src/ripple/consensus/Validations.h

```diff
@@ -587,11 +603,46 @@ class Validations
             std::lock_guard lock{mutex_};
 
             // Check that validation sequence is greater than any non-expired
-            // validations sequence from that validator
+            // validations sequence from that validator; if it's not, perform
+            // additional work to detect Byzantine validations
             auto const now = byLedger_.clock().now();
-            SeqEnforcer<Seq>& enforcer = seqEnforcers_[nodeID];
-            if (!enforcer(now, val.seq(), parms_))
+
+            auto const [seqit, seqinserted] = bySequence_[val.seq()].emplace(nodeID, val);
+
+            if (!seqinserted)
+            {
+                // Check if the entry we're already tracking was signed
+                // long enough ago that we can disregard it.
+                auto const diff =
+                    std::max(seqit->second.signTime(), val.signTime()) -
+                    std::min(seqit->second.signTime(), val.signTime());
+
+                if (diff > parms_.validationCURRENT_WALL &&
+                    val.signTime() > seqit->second.signTime())
+                    seqit->second = val;
+            }
+
+            // Enforce monotonically increasing sequences for validations
+            // by a given node:
+            if (auto& enf = seqEnforcers_[nodeID]; !enf(now, val.seq(), parms_))
+            {
+                // If the ledger hash matches, it must be a validation for
+                // the same ledger:
+                bool const same = (seqit->second.ledgerID() == val.ledgerID());
+
+                // Two validations for the same ledger but with different
+                // cookies. This is probably accidental misconfiguration.
+                if (same && seqit->second.cookie() != val.cookie())
+                    return ValStatus::multiple;
+
+                // Two validations for the same sequence but for different
+                // ledgers. This could be the result of misconfiguration
+                // but it can also mean a Byzantine validator.
+                if (!same)
```

Comment:
If they have different ledger IDs, what guarantees they still have the same sequence number at this spot in the code? Do you also have to check `seqInserted`?

Reply:
I'll double-check this. Thanks.